### ISSUE TITLE
Poor detections with SwinB backbone

### ISSUE THREAD

**Issue by danigarciaoca** [IDEA-Research/GroundingDINO#128]
Hi everyone,

I was testing Grounding Dino with SwinB backbone and I found that, unexpectedly, it detects less objects than Grounding Dino SwinT. For example, when asking for "light bulb", G-DINO SwinT finds almost all bulbs but G-DINO SwinB founds no one. Any explanation?

```
IMAGE_PATH = "weights/dog-3.jpeg"
TEXT_PROMPT = "light bulb"
BOX_TRESHOLD = 0.35
TEXT_TRESHOLD = 0.25
```

**G-DINO SwinT (backbone = "swin_T_224_1k"):**
![annotated_image_1](https://github.com/IDEA-Research/GroundingDINO/assets/27818263/6773f64f-4ff0-47fe-b5c9-af5ed49f7a85)

**G-DINO SwinB (backbone = "swin_B_384_22k")**
![annotated_image_2](https://github.com/IDEA-Research/GroundingDINO/assets/27818263/f4a08952-b00b-4042-b37b-7b7058e28e50)

I also tried tunning thresholds but it randomly detects other objects. Any idea? Am I missing any configuration detail?

Thanks in advance!

**Comment by rentainhe** [comment]
Hello, do you use the same `box_threshold` for both Swin-T and Swin-B, you can set a lower threshold for Swin-B model

**Comment by dankresio** [comment]
Hi! Yes, I also tried lowering the threshold and a lot of false positive are triggered. Some examples:

`BOX_TRESHOLD = 0.10`:
![annotated_image](https://github.com/IDEA-Research/GroundingDINO/assets/27818263/c00a9d54-6853-4c57-a52d-3a5707482e82)

`BOX_TRESHOLD = 0.15`
![annotated_image](https://github.com/IDEA-Research/GroundingDINO/assets/27818263/f552b839-038c-40f6-bc7a-9381d8886f20)

**Comment by rohit901** [comment]
Hi @dankresio 
I tried the swin_B model on your image and the default threshold values and I'm getting the same results as you. It is not able to detect light bulb. I even tried changing the prompt to just "bulb", that also did not work. However it is able to detect other classes like dog, sunglasses.
`TEXT_PROMPT = "dog . bulb . sunglasses ."`
![annotated_image2](https://github.com/IDEA-Research/GroundingDINO/assets/30185369/45d5ebc0-1e2b-4bd9-9228-7d22e9c5d27a)

**Comment by psykokwak-com** [comment]
Same here.
I tested the two models on thousands of images and swin_T works better than swin_B

**Comment by dankresio** [comment]
Hi @rohit901. The problem is that:

1. Even if it can detect some other classes like "dog" or "sunglasses" correctly, they are detected with lower confidence when using SwinB than when using SwinT (however, confidence changes could be caused due to backbone architecture changes) and some others are missing.
2. SwinB model is supposed to be more powerful thant SwinT

Just to give some context:

- SwinT (tiny) backbone has 86M parameters and has been trained with 1k classes and input image size 224, while
- SwinB (base) backbone has 145M parameters and has been trained with 22k classes and input image size 384

When increasing architecture size and the number of classes used for pre-training the backbone, you let the backbone learn richer features, which leads to better performance
When increasing input image size, you let the model to improve spatial resolution and so, detect smaller objects

So, to sum up, rationally it does not make sense that "G-DINO SwinB" flavour works worse than "G-DINO SwinT".

Btw, @psykokwak-com thanks for corroborating the issue!

@rentainhe @HaoLiuHust , is there any idea of releasing SwinL backbone for G-DINO? or an improved version of SwinB backbone.

Thanks in advance, and great work with this repo! I'm enjoying a lot testing it

**Comment by laisimiao** [comment]
I also have found this problem after testing some pictures

**Comment by rohit901** [comment]
Guys did anyone of you were able to replicate the results mentioned in the paper on the LVIS or LVIS minival dataset?? Unfortunately I'm not able to replicate those results and I'm getting very poor AP on LVIS when trying to evaluate with SwinB.. is it because of my Box/Text thresholds? I'm using 0.32 Box and 0.3 Text thresholds.. Splitting 1203 LVIS classes into length of 125 each to pass as text prompt due to max token limit of BERT.

passing the batched data as follows
```
image = image.repeat(len(TEXT_PROMPT_LIST), 1, 1, 1) # repeat same image in first dimension to cover all lvis classes. batch dim covers all lvis classes
with torch.no_grad():
     output = model(image, captions = TEXT_PROMPT_LIST)
```

**Comment by dankresio** [comment]
@rohit901 I think that the point is that Swin-T weights are fine tuned in COCO while Swin-B weights not (just my conclusion, doesn't have to be the correct one)

**Comment by rohit901** [comment]
@danigarciaoca, I wanted to replicate the results mentioned in Table 3 of the paper on LVIS, where I think even Swin-T is getting AP of 25.6 / 27.4.. from my experiments I was not able to replicate that AP.. 

I will be checking their latest code on zero-shot COCO eval and try to report back.

**Comment by danigarciaoca** [comment]
I'll stay tuned! thanks @rohit901

**Comment by rohit901** [comment]
I'm discussing the mAP result replication in this separate thread https://github.com/IDEA-Research/GroundingDINO/issues/147 if you're interested to follow on @danigarciaoca..

It seems we aren't supposed to be filtering out low confidence/noisy box prediction with thresholds when calculating mAP metric, it is only for visualization purposes I guess?

**Comment by aixiaodewugege** [comment]
> @rohit901 I think the main point is that Swin-T weights are fine tuned in COCO while Swin-B weights are not (just my conclusion, doesn't have to be the correct one)

I think  Swin-B weights are finetuned in COCO, so its open-set detection ability is heavily inferenced, for COCO only have 80 class. That is why you can detect dog and glasses well with Swin-B weights rather than light bulb~~

**Comment by rohit901** [comment]
> > @rohit901 I think the main point is that Swin-T weights are fine tuned in COCO while Swin-B weights are not (just my conclusion, doesn't have to be the correct one)
> 
> I think Swin-B weights are finetuned in COCO, so its open-set detection ability is heavily inferenced, for COCO only have 80 class. That is why you can detect dog and glasses well with Swin-B weights rather than light bulb~~

the authors @rentainhe @SlongLiu can confirm

**Comment by DianCh** [comment]
May I ask where you guys downloaded the SwinB weight? The instruction in README only provided this link:

```
mkdir weights
cd weights
wget -q https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha/groundingdino_swint_ogc.pth
cd ..
```

**Comment by iamnotthatbob** [comment]
> May I ask where you guys downloaded the SwinB weight? The instruction in README only provided this link:
> 
> ```
> mkdir weights
> cd weights
> wget -q https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha/groundingdino_swint_ogc.pth
> cd ..
> ```

It's in the latest [release](https://github.com/IDEA-Research/GroundingDINO/releases/tag/v0.1.0-alpha2) (i.e. v0.1.0-alpha2). The direct link being: `https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha2/groundingdino_swinb_cogcoor.pth`.

Also, you would need to use [this config file](https://github.com/IDEA-Research/GroundingDINO/blob/main/groundingdino/config/GroundingDINO_SwinB_cfg.py) instead of the one used in most tutorials with SwinT.

**Comment by TNPIsHere** [comment]
Hey, I know, it's been a while since topic started, but I see the same. Wanted to use GDINO for unlabeled dataset annotation. This is what I got, just fyi. Have no idea why though. I used standard 0.35, 0.25 for thresholds, and "license plate" as a prompt. SwinT - almost perfect (even though it is because the photo is of a quite good quality and stuff), while SwinB - complete miss, instead it detects the whole vehicle.

<img width="454" height="310" alt="Image" src="https://github.com/user-attachments/assets/14efd750-a13d-41ac-bcc9-e1d52558642d" />

<img width="468" height="309" alt="Image" src="https://github.com/user-attachments/assets/f6b25bf6-00b3-4437-8694-684b37d00b97" />

UPDATE: I asked LLM and I really liked one specific point it output. As SwinB is a "heavier" model, it may generalize worse, and have issues with identifying very small objects on an image. So it might be better at detecting large objects than T, but performs worse on small objects (light bulb as well as car license plates are small objects relative to the image size).

Another good point was that it processes text prompts differently: what worked well for "T" might completely fail for "B". I noticed that B poorly identifies "license plate", but performs comparable (or better) to T with the usage of "vehicle registration plate". Play around with the prompt.
